account.go

```diff
@@ -88,11 +88,19 @@ func (*ACMEIssuer) newAccount(email string) (acme.Account, error) {
 // If it does not exist in storage, it will be retrieved from the ACME server and added to storage.
 // The account must already exist; it does not create a new account.
 func (am *ACMEIssuer) GetAccount(ctx context.Context, privateKeyPEM []byte) (acme.Account, error) {
-	account, err := am.loadAccountByKey(ctx, privateKeyPEM)
-	if errors.Is(err, fs.ErrNotExist) {
-		account, err = am.lookUpAccount(ctx, privateKeyPEM)
+	email := am.getEmail()
+	if email == "" {
+		account, err := am.loadAccountByKey(ctx, privateKeyPEM)
+		if err == nil || !errors.Is(err, fs.ErrNotExist) {
```

Comment:
`loadAccountByKey` can return different error types: "whatever storage returned for a non-existing key" and `fs.ErrNotExist` when there is indeed no account at all.

There doesn't seem to be a requirement for the storage to return fs error codes, and it also seems a bit odd that it aborts the first time and not `continue`s to the end of the list of accounts:
https://github.com/caddyserver/certmagic/blob/f64401c80d50d396f1cedefb316108c10ebcbf84/account.go#L108-L111

WDYT: Would a `continue` there make more sense, and then in `GetAccount` we can just drop the `!errors.Is(err, fs.ErrNotExist)` half and always look up the account if one wasn't found by `loadAccountByKey`?

Reply:
> loadAccountByKey can return different error types: "whatever storage returned for a non-existing key" and fs.ErrNotExist when there is indeed no account at all.

Hmm, I would expect the storage implementation to return `fs.ErrNotExist` whenever the key or account does not exist. If the key doesn't exist, that means the account doesn't exist (in storage). Thus it needs to be looked up. (Maybe I'm missing something though)

> There doesn't seem to be a requirement for the storage to return fs error codes,

Maybe I should clarify that in the docs. My understanding though is that `fs` errors are conventional now, and Go std lib uses and recommends them all over the place (os, io packages for example). We can clarify that in the storage API godocs if needed. :+1:

> WDYT: Would a continue there make more sense, and then in GetAccount we can just drop the !errors.Is(err, fs.ErrNotExist) half and always look up the account if one wasn't found by loadAccountByKey?

Yes I suppose that would make sense, but I think an error there implies a broken storage backend or a corrupted file structure (i.e. a file that does not exist but should). So I dunno... MAYBE ignoring is the right thing to do there... ?

Reply:
If the storage is corrupted indeed it could be a different problem, so continuing could also be stupid. Or you're lucky, and the thing that's corrupted wasn't what you were looking for anyways, and the thing you _are_ looking for is the next on the list.

The storage listing said there would be a prefix for that email, but that doesn't say anything about the files inside that prefix. If something failed to write the key file into that prefix earlier, we would now just fail -- regardless whether this is the right thing in the first place. IOW: The storage corruption in this case isn't "disk is failing" but "_we_ earlier failed". Then again, if you do fast-exit here, we go into "lookup via key from ACME", and assuming that key was correct then we will now retry to save the account -- regardless whether that was indeed this broken key (possibly healing the earlier problem) or another already existing key (which we could have found, had we not stopped). 

`continue`-ing here avoids the second case problem: If what is broken is not what we're looking for, and what we're looking for does exist, we will find it, and avoid the ACME call + redundant save.

Reply:
I'm convinced about the continue. That's probably right. Thanks for hashing this out!

Reply:
Going to add it as a second commit to this PR then :)

Reply:
https://github.com/caddyserver/certmagic/pull/283/commits/97a23b904c7e6dcf820b4aa381eebb1ab8873ef3